openwisp_controller/api/serializers.py

```diff
@@ -0,0 +1,137 @@
+        allow_null=False, queryset=Organization.objects.all(), required=True
+    )
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(FilterSerializerByOrgOwned, ValidatedModelSerializer):
+    config = serializers.JSONField()
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = '__all__'
+
+
+class DeviceConfigSerializer(serializers.ModelSerializer):
+    config = serializers.JSONField()
+    context = serializers.JSONField()
+
+    class Meta:
+        model = Config
+        fields = ['backend', 'status', 'templates', 'context', 'config']
+
+
+class DeviceListSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer(write_only=True)
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = [
+            'id',
+            'name',
+            'organization',
+            'mac_address',
+            'key',
+            'last_ip',
+            'management_ip',
+            'model',
+            'os',
+            'system',
+            'notes',
+            'config',
+        ]
+
+    def create(self, validated_data):
+        config_data = validated_data.pop('config')
+        device = Device.objects.create(**validated_data)
+        config_templates = config_data.pop('templates')
+        config = Config.objects.create(device=device, **config_data)
+        for template in config_templates:
+            config.templates.add(template.pk)
+        return device
+
+
+class DeviceDetailSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer()
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = DeviceListSerializer.Meta.fields
+
+    def update(self, instance, validated_data):
+        config_data = None
+
+        if self.initial_data.get('config.backend') and instance.backend is None:
+            new_config_data = dict(validated_data.pop('config'))
+            config_templates = new_config_data.pop('templates')
+            config = Config.objects.create(device=instance, **new_config_data)
+            for template in config_templates:
+                config.templates.add(template.pk)
+            return super().update(instance, validated_data)
+
+        if validated_data.get('config'):
+            config_data = validated_data.pop('config')
+
+        if config_data:
+            instance.config.backend = config_data.get(
+                'backend', instance.config.backend
+            )
+            instance.config.status = config_data.get('status', instance.config.status)
+
+            config_templates = config_data.get('templates')
+            instance.config.templates.clear()
```

Comment:
Please do not clear templates by default. Only do this if there is any type of change to the templates.
Ensure you have current templates as a list of strings with UUIDs, and the list of templates coming from the HTTP request also as a list of strings with UUID, then only if they're different the templates should be cleared and added back.

Reply:
Done